File: app_codebase/dart_bkend_base_02/bin/core/ws/contracts/register_channel_contract.dart
/// Let feature modules register WebSocket channel handlers without knowing the registry.
library;

import 'ws_message_contract.dart';

abstract interface class ApplicationChannelSink {
  void publicChannel(String channel, WsChannelHandler handler);
  void authuserChannel(String channel, WsChannelHandler handler);
  void serviceChannel(String channel, WsChannelHandler handler);
}

File: app_codebase/python_base_05/bin/core/ws/contracts/register_channel_contract.py
"""Let feature modules register WebSocket channel handlers without knowing the registry."""

from __future__ import annotations

from typing import Any, Callable, Protocol

from core.ws.contracts.ws_message_contract import WsClientMessage, WsConnectionContext

WsChannelHandler = Callable[[WsConnectionContext, WsClientMessage], dict[str, Any] | None]


class ApplicationChannelSink(Protocol):
    """Register channel handlers per tier.

    **Channel** names are logical (e.g. ``demo/echo``) — do not include ``/authuser`` prefix.
    """

    def public_channel(self, channel: str, handler: WsChannelHandler) -> None: ...

    def authuser_channel(self, channel: str, handler: WsChannelHandler) -> None: ...

    def service_channel(self, channel: str, handler: WsChannelHandler) -> None: ...
